UIP-625: added TextInput component

rey-wright opened the pull request:
## Description
Added a TextInput component

## Screenshots
<img width="1056" alt="Screen Shot 2020-02-25 at 12 17 17 PM" src="https://user-images.githubusercontent.com/52427513/75270576-c8133880-57c8-11ea-8c96-5c5e3d973ffd.png">

## Checklist
- [x] Docs have been rebuilt (`yarn build:full`)
- [x] All unit tests pass
- [x] Snapshots have been updated
- [x] No lint errors or warnings have been introduced
- [x] **[Version bumped](https://github.com/cbinsights/form-design-system#updating-version-number) if appropriate**

rey-wright commented:
@akdetrick in reference to our 1:1 today, I'm happy to remove the stories for IconInput and DecoratedInput if that feels like a good move to make once you look through this PR.

rey-wright commented on src/components/forms/TextInput.jsx:
```diff
@@ -0,0 +1,115 @@
+}) => {
+  // Refer to the following for an explanation on
+  // the throttling / debouncing approach:
+  // https://www.peterbe.com/plog/how-to-throttle-and-debounce-an-autocomplete-input-in-react
```
I highly highly recommend reading this article, just to make sure we're aligned on the approach. There may be different approaches / articles we can look into as well.

There's actually another thing in the article that I might steal... throttling if the length of the characters is less than 5, AND if the ending character ever ends in a space. Article explains why you'd want to do that.

akdetrick commented on CHANGELOG.md:
```diff
@@ -27,6 +27,10 @@ This is an example of a brief overview of the _Major_ or _Minor_ version changes
 
 ---
 
+## [7.8] Add TextInput
```
you'll need to bump this again (merge from master)

akdetrick commented on src/components/forms/TextInput.jsx:
```diff
@@ -0,0 +1,115 @@
+  errorText: PropTypes.string,
+
+  /** Label used for `label` element */
+  label: PropTypes.string,
```
How do you feel about making this required and having a `showLabel` prop the way we do with current checkbox?

If a developer wants to hide the label, we use the `label` string on an `aria-label` attribute of the input.

rey-wright replied:
So yeah I guess this becomes a bit tricky. I can see pros and cons of both sides of this.

If we force the developer to pass a label regardless, then what happens to the visual label that they need to show users? Do they still use their own label (so that there are 2 labels)? Or do they just add a span with some text, and the "true" label is on the `aria-label`? I sort of see the benefit of making label `required` so that devs don't abuse not passing a label (which will 100% happen), but then again I sort of question the developer having to apply "2" labels, the one they pass to the component, and then the one they specify themselves, and then the whole thing where they click on their own label... would that focus the input. Lot of... variables here 🙂

akdetrick replied:
The pattern would be...

```
label: PropTypes.string.isRequired,
showLabel: PropTypes.bool, // default `true` - if set to `false`, apply `label` to `aria-label`
```

rey-wright replied:
right but I don't get the surrounding things that happen as a result of this. We should maybe chat about it, it's slightly confusing.

akdetrick commented on src/components/forms/TextInput.jsx:
```diff
@@ -0,0 +1,115 @@
+   */
+  IconRight: PropTypes.func,
+
+  /** Standard input type */
```
minor suggestion

```suggestion
  /** Input element `type` attribute */
```

akdetrick commented on src/components/style/forms/decoratedInput.css:
```diff
@@ -29,7 +29,8 @@
   /* use line box to center text and match input height */
   padding: 0 var(--space-half);
-  line-height: var(--space-double);
+  display: inline-flex;
```
🙏

akdetrick commented on src/components/forms/textInput.stories.mdx:
```diff
@@ -0,0 +1,177 @@
+
+### Misc
+
+Can be used with InputGroup (why don't these stretch full width?)
```
was this an unaddressed TODO?

rey-wright replied:
This is possibly more of a question for you I think? IconInput, DecoratedInput, etc, all stretch to be 100% width, except InputGroup. Sort of wondering if it's expected that this is different, or if it should be the same.

I'm glad I wrote this because I completely forgot about it.

akdetrick replied:
the outer border for an input group is the root element of the `InputGroup` component. This element is set to `min-content` because the last `input` or `select` element in the group does not grow to fill remaining space.

akdetrick commented on src/components/forms/TextInput.jsx:
```diff
@@ -0,0 +1,115 @@
+
+export const throttleValue = 500;
+
+const TextInput = ({
```
Needs unit tests

rey-wright replied:
Hmmm so I can do snapshot tests as an easy fix. I could "also" test the throttling functions but I'm hesitant to test that, unless we're confident that we don't want to change the throttling configuration.

akdetrick replied:
Snapshots to cover different prop combinations would work. For throttling, maybe spy on `onThrottledChange` - Jest timers may be useful here. If it's too much of a pain, snapshots might be enough.

rey-wright replied:
I'm now testing that onThrottldeChange is now firing. I'm opting to test the function I'm using directly to test whether the debounce vs throttle variant of the function fires.

akdetrick commented on src/components/forms/TextInput.jsx:
```diff
@@ -0,0 +1,115 @@
+
+export const throttleValue = 500;
+
+const TextInput = ({
```
Should we remove the stories for `IconInput` and `DecoratedInput` in this branch?

rey-wright replied:
I'm down, as long as you agree / see this component replacing `IconInput` / `DecoratedInput`.

rey-wright replied:
I'll take this question as an implicit thumbs up and go ahead and remove it for now 🙂

akdetrick replied:
Yep, I'd vote on removing those from docs

rey-wright commented on src/components/forms/TextInput.jsx:
```diff
@@ -0,0 +1,115 @@
+  label,
+  errorText,
+  throttleOnChange,
+  required,
```
Interesting... 

So what I've been doing personally, is if the boolean is spread directly onto the component as expected, even if it causes additional side effects, I keep it that name. So like... a perfect example is the disabled attribute of button. Technically, anchors don't have a disabled prop, but I figured it made more sense to accommodate for the `button` element that does.

In this change, if input itself uses the required prop as it normally does, I'd at least "historically" choose to keep calling it required, so that the developer knows that the prop carries the same intention as what the normal input would. That being "said", I can see the alternative angle of using the `is` prefix the moment it's being used in any capacity beyond its normal usage.

Essentially, I don't feel our usage guidelines totally encapsulates this specific use case.

All that being said, that's just what I've done historically, and honestly either way would make sense to me. What do you think makes the most sense in this situation?

rey-wright replied:
The other side thing to note: funny that you mention it, but I technically did "not" add the required field to the input yet. Are we comfortable with the UI that makes the input border completely red if it's a required field? Because that's what happens to our current input.

rey-wright replied:
<img width="1115" alt="Screen Shot 2020-03-02 at 1 15 24 PM" src="https://user-images.githubusercontent.com/52427513/75704717-e9bc6600-5c87-11ea-9edd-514b210ddc5f.png">

<img width="1703" alt="Screen Shot 2020-03-02 at 1 16 00 PM" src="https://user-images.githubusercontent.com/52427513/75704769-ffca2680-5c87-11ea-869f-0af18a45f693.png">

So if required is applied to input, this is how it would look.

akdetrick replied:
I suspect that this is the `:invalid` state (required but empty). Does it only happen on blur? If so, it might be okay...

rey-wright replied:
@akdetrick unfortunately I believe this happens regardless of blur.

rey-wright commented on src/components/forms/TextInput.jsx:
```diff
@@ -0,0 +1,116 @@
+const TextInput = ({
+  label,
+  errorText,
+  throttleOnChange,
```
@akdetrick how do you feel about `onThrottleChange`?

akdetrick replied:
`onThrottledChange` makes the most sense to me (past tense)

akdetrick replied:
`onThrottledChange` (past tense Throttled) makes the most sense to me

akdetrick commented on src/components/forms/textInput.stories.mdx:
```diff
@@ -0,0 +1,224 @@
+
+<Story name="Controlled Input">
+  <TextInput
+    value="Controlled"
```
our example should also show an `onChange` prop for controlled input

akdetrick commented on src/components/forms/textInput.stories.mdx:
```diff
@@ -0,0 +1,224 @@
+</Story>
+
+### Form management library
+You should be able to use this input in any form management library you'd like:
```
I'd like to declare our preference for react-hook-form here

```suggestion
You may use this input in any form management library you'd like. We recommend `react-hook-form`, which is used in the example below:
```

rey-wright replied:
thanks for the suggested change 👍

rey-wright commented on CHANGELOG.md:
```diff
@@ -27,6 +27,10 @@ This is an example of a brief overview of the _Major_ or _Minor_ version changes
 
 ---
 
+## [7.10] Add TextInput
```
no worries 👍 

In general I usually wait on solving conflicts like these until after approval, as I'll just need to solve them again if not.

rey-wright commented:
@akdetrick one note here: I added support for "orientation" if you spin up the server and take a look at the knobs. I "may" need some of your flex expertise to get it working properly, but this can always be addressed later via a patch / minor update.

akdetrick commented on src/components/forms/TextInput.jsx:
```diff
@@ -0,0 +1,143 @@
+  /** String to place to the right of the input */
+  after: PropTypes.string,
+
+  orientation: PropTypes.oneOf(['vertical', 'horizontal']),
```
Only thing left to address here - I suspect the `horizontal` orientation may more commonly need the label on the left of the input.

This also opens up a potential API change if we need to give users control over left/right/top

rey-wright replied:
@akdetrick added!

rey-wright commented:
@akdetrick Oh I forgot to ask my 2nd question: I added some interesting flex stuff to this PR to accommodate for the orientation changes. Wondering if you could take a look at the flex CSS I added to see if it makes sense (and if there's any chance adding those changes could break anything). This is mostly around changing CSS for things like fdsDecoratedInput and the like.

rey-wright commented:
I'll resolve conflicts on this when it gets approved.

rey-wright commented:
@akdetrick 

- `orientation` is now `labelPosotion`
- `left` and `top` are the only valid values now. `right` has been removed.
